src/facade.ts

```diff
@@ -0,0 +1,39 @@
+import type * as AppMap from "./AppMap";
+
+import { recording, start } from "./recorder";
+import {
+  disableGlobalRecording,
+  startCodeBlockRecording,
+  stopCodeBlockRecording,
+} from "./recorderControl";
+import Recording from "./Recording";
+
+// Since _this_ module is loaded, we'll do code block recording only.
+recording.abandon();
+disableGlobalRecording();
+
+export function record(block: () => unknown): AppMap.AppMap | undefined {
+  start(new Recording("block", "block", new Date().toISOString()));
+  startCodeBlockRecording();
+  try {
+    block();
+  } finally {
+    stopCodeBlockRecording();
+    recording.finish();
+  }
+  return recording.readAppMap();
+}
+
+export async function recordAsync(block: () => unknown): Promise<AppMap.AppMap | undefined> {
+  start(new Recording("block", "block", new Date().toISOString()));
+  startCodeBlockRecording();
+  try {
+    await block();
+  } finally {
+    stopCodeBlockRecording();
+    recording.finish();
+  }
+  return recording.readAppMap();
+}
```

Comment:
> My last decision was to make separate record functions for sync and async usages. I thought this was the best option for the client code. Otherwise the return type of the record function had to be `Promise<AppMap.AppMap | undefined> | AppMap.AppMap | undefined`, complicating consumption of the result by client code. Clients would need to make type assertions in Typescript to get intellisense etc. Now they can access members easilly: `appmap?.events.length`.

You can overload the declaration:

```typescript
type NotPromise<T> = T extends Promise<unknown> ? never : T;

export function record<T>(block: () => NotPromise<T>): AppMap.AppMap | undefined;
export function record<T>(block: () => Promise<T>): Promise<AppMap.AppMap> | undefined;

export function record(block: () => unknown): AppMap.AppMap | Promise<AppMap.AppMap> | undefined {
  start(new Recording("block", "block", new Date().toISOString()));
  startCodeBlockRecording();
  try {
    const result = block();
    if (result instanceof Promise) return result.then(finishRecording, finishRecording);
  } catch {
    // intentionally empty
  }
  return finishRecording();
}

function finishRecording(): AppMap.AppMap | undefined {
  stopCodeBlockRecording();
  if (!recording.finish()) return;
  return recording.readAppMap();
}
```